### ISSUE TITLE
Error on Building the Repository

### ISSUE THREAD

**Issue by Kahbazi** [dotnet/aspnetcore#21941]
I can't build the repository. I'm getting this error.

> E:\Github\Repos\Kahbazi\AspNetCore\.dotnet\sdk\5.0.100-preview.5.20258.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.Packa
> geDependencyResolution.targets(234,5): error NETSDK1005: Assets file 'E:\Github\Repos\Kahbazi\AspNetCore\artifacts\obj\
> RepoTasks\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v5.0'. Ensure that restore has run and th
> at you have included 'net5.0' in the TargetFrameworks for your project. [E:\Github\Repos\Kahbazi\AspNetCore\eng\tools\R
> epoTasks\RepoTasks.csproj]

I checked the `artifacts\obj\RepoTasks\project.assets.json` and I saw `.NETFramework,Version=v5.0` in the `targets`. Is this correct? Do we have `.NET**Framework**,Version=v5.0``?
Is there anything I can do to fix the build?

**Comment by Tratcher** [comment]
There was just a major tooling update to move to net5.0 TFMs, you may need to `git clean -xdf` and rebuild. I was able to build on the current master (b1fe320).

**Comment by ilkayilknur** [comment]
I'm getting the same error while building the repository.

**Comment by Kahbazi** [comment]
@Tratcher I got the same error with the latest commit. Do you also have `.NETFramework,Version=v5.0`  in `artifacts\obj\RepoTasks\project.assets.json`? Shouldn't this be `.NETCoreApp,Version=v5.0`?

**Comment by adamsitnik** [comment]
I am also getting the same error:

```log
PS C:\Projects\aspnetcore> git clean -xfd
Removing .dotnet/
Removing .tools/
Removing artifacts/
PS C:\Projects\aspnetcore> .\build.cmd -c Release
WARNING: No default group of projects was specified, so building the 'managed' and its dependent subsets of projects. Run `build.cmd -help` for more details.
Building of NodeJS projects is disabled since node is not detected on Path and no BuildNodeJs or NoBuildNodeJs setting is set explicitly.
WARNING: Some managed projects depend on NodeJS projects. Building NodeJS is disabled so the managed projects will fallback to using the output from previous builds. The
output may not be correct or up to date.
Detected JDK 1.8 in C:\Program Files\Java\jdk1.8.0_251 (via registry)
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.5.20258.4/dotnet-sdk-5.0.100-preview.5.20258.4-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.5.20258.4/dotnet-sdk-5.0.100-preview.5.20258.4-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\Projects\aspnetcore\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
Downloading vswhere
Downloading RoslynTools.MSBuild 16.5.0-alpha
  Tool 'dotnet-serve' (version '1.5.0') was restored. Available commands: dotnet-serve

  Restore was successful.
  Restore completed in 412.2 ms for C:\Users\adsitnik\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20261.9\tools\Tools.proj.
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.6.20264.1/dotnet-runtime-5.0.0-preview.6.20264.1-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.6.20264.1/dotnet-runtime-5.0.0-preview.6.20264.1-win-x64.zip
dotnet-install: Installation finished
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.6.20264.1/dotnet-runtime-5.0.0-preview.6.20264.1-win-x86.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.6.20264.1/dotnet-runtime-5.0.0-preview.6.20264.1-win-x86.zip
dotnet-install: Adding to current process PATH: "C:\Projects\aspnetcore\.dotnet\x86\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
  Restore completed in 1.1 sec for C:\Projects\aspnetcore\src\Analyzers\Internal.AspNetCore.Analyzers\src\Internal.AspNetCore.Analyzers.csproj.
  Restore completed in 1.12 sec for C:\Projects\aspnetcore\eng\tools\RepoTasks\RepoTasks.csproj.
C:\Projects\aspnetcore\.dotnet\sdk\5.0.100-preview.5.20258.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(234,5): error NETSDK1005: Assets f
ile 'C:\Projects\aspnetcore\artifacts\obj\RepoTasks\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v5.0'. Ensure that restore has run and that you have
 included 'net5.0' in the TargetFrameworks for your project. [C:\Projects\aspnetcore\eng\tools\RepoTasks\RepoTasks.csproj]
```

**Comment by ilkayilknur** [comment]
I can build the repository on Mac but it fails on Windows.

**Comment by Tratcher** [comment]
@dougbu ?

**Comment by dougbu** [comment]
We saw this error when running our build pipeline on agents with an older version of Visual Studio. The root cause is that version ships with an `msbuild` that does not understand the `net5.0` TFM. Visual Studio 2019 16.6 preview 2 or later.

I recommend you all upgrade your Visual Studio installations. 16.6 RTM is now available for download: https://visualstudio.microsoft.com/downloads/

**Comment by dougbu** [comment]
I just merged a change to our global.json (58cf2304a642) which should make the error message more clear if your VS installation is too old to build this repo.

Unfortunately, missing VS components may also lead to similar problems. See https://github.com/dotnet/aspnetcore/pull/21958#issuecomment-630923196

@Kahbazi @adamsitnik @ilkayilknur please let us know if upgrading VS and double-checking the installed components unblocks you. If not, please reopen this issue.

**Comment by ilkayilknur** [comment]
I upgraded VS to 16.6. I'm getting a different error while executing restore command. 

```
PS C:\source\aspnetcore> .\restore.cmd
Building of NodeJS projects is disabled since node is not detected on Path and no BuildNodeJs or NoBuildNodeJs setting is set explicitly.
WARNING: Some managed projects depend on NodeJS projects. Building NodeJS is disabled so the managed projects will
fallback to using the output from previous builds. The output may not be correct or up to date.
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.5.20258.4/dotnet-sdk-5.0.100-preview.5.20258.4-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.5.20258.4/dotnet-sdk-5.0.100-preview.5.20258.4-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\source\aspnetcore\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
Downloading vswhere
Downloading RoslynTools.MSBuild 16.6.0-alpha
No package matching package identifier RoslynTools.MSBuild and version 16.6.0-alpha was found in the given feed.
```

Any ideas ? @dougbu

**Comment by dougbu** [comment]
@ilyas1974 that error occurs sometimes when eng/common/tools.ps1 can't find a VS installation meeting the requirements. It's attempting a fallback download of a package version that apparently doesn't exist. But, you shouldn't need the fallback unless your installation is incomplete.

That script executes something like
``` powershell
& 'C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe' -version 16.6 `
-latest -prerelease -products * `
-requires Microsoft.Component.MSBuild `
-requires Microsoft.VisualStudio.Component.VC.ATL `
-requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 `
-requires Microsoft.VisualStudio.Component.Windows10SDK.17134
```

Try that out then eliminate `-requires ....` options to find what you're missing.

**Comment by ilkayilknur** [comment]
Installing Windows10SDK.17134 fixed the issue. Thanks for the help @dougbu

**Comment by adamsitnik** [comment]
I am getting the same error as @ilkayilknur 

![obraz](https://user-images.githubusercontent.com/6011991/82438251-0737e980-9a99-11ea-87b3-c94b3901d459.png)


I am going to try to update the Windows SDK and see if it helps (BTW why does ASP.NET need an OS SDK?)

**Comment by adamsitnik** [comment]
Installing the SDK helped. Thanks!

**Comment by davidfowl** [comment]
I get this same issue.

**Comment by dougbu** [comment]
@davidfowl did you figure out what the missing VS component was?

**Comment by davidfowl** [comment]
Its still broken for me. @AdamBraden what did you have to install.

**Comment by BrennanConroy** [comment]
I updated from a version of VS 16.5 to 16.7 and it fixed the issue for me.

**Comment by dougbu** [comment]
@BrennanConroy that's overkill 😃 since only 16.6 is needed. Then again, I'm still working to make 16.5 the minimum again.

**Comment by davidfowl** [comment]
This experience isn't good. Where do these requirements come from and why can't I run a single thing to do it automagically to get me in a good state?

**Comment by davidfowl** [comment]
https://github.com/dotnet/aspnetcore/issues/21941#issuecomment-631376584 this worked for me too...

**Comment by dougbu** [comment]
@davidfowl we don't usually need to change our VS requirements though we've been tracking RTM versions lately. However, the `net5.0` change forced a move to what was then a preview version.

One thing that can _usually_ get you into a good state is
``` powershell
.\eng\scripts\InstallVisualStudio.ps1 -Channel Release -Edition Professional -InstallPath 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Release'
```
I don't usually suggest this because the command-line arguments are easy to get wrong.

----

I agree the Arcade fallback is broken in that it's attempting to download a `msbuild` version that can't be found. Reading eng/common/tools.ps1 closely, it appears I can change global.json to at least disable the unwanted fallback (the `xcopy` version of `msbuild` is unlikely to be sufficient). I'll try this out sometime today or tomorrow.

**Comment by dougbu** [comment]
#22103

**Comment by dougbu** [comment]
Resolution didn't work because #22103 broke official builds. Reverting that change in #22166.

@epananth could you please file a separate issue somewhere (not sure of the appropriate repo) to allow global.json to specify where the "xcopy-msbuild" configuration can be used? Falling back to VS 2019 16.5 or a desktop `msbuild` that old will break local developers and cause strange errors -- as seen above. What I'd like to specify is that the fallback is _not_ attempted except in the post-build steps or (maybe) only in CI builds.

**Comment by dougbu** [comment]
Looks like I forgot to follow up on this one. The aspnetcore repo again supports building with VisualStudio 2019 16.5. People may still want to update to the now-RTM 16.6 version and may still hit issues when their 16.6 installation is missing some required components. For those cases, the InstallVisualStudio.ps1 fix mentioned in https://github.com/dotnet/aspnetcore/issues/21941#issuecomment-632228104 is probably the best way to go.
